File: pkg/quantize/ts/prediction-buffer.ts
import { isForwardSequence, sequenceDistance } from "./playback-controller.js";

export const DEFAULT_MAX_PENDING_INPUTS = 128;

export interface PendingInput<TInput> {
  seq: number;
  input: TInput;
}

export interface PredictionBufferConfig {
  maxPending?: number;
}

export interface PushInputResult<TInput> {
  accepted: boolean;
  /** Oldest unacknowledged input evicted to keep the buffer bounded. */
  dropped?: PendingInput<TInput>;
  reason?: "duplicate-or-stale";
}

export interface AcknowledgeResult {
  advanced: boolean;
  acknowledgedCount: number;
  pendingCount: number;
}

export interface ReconciliationResult<TState> {
  state: TState;
  pendingCount: number;
  acknowledgedCount: number;
}

function isAcknowledgedBy(sequence: number, cumulativeAck: number): boolean {
  const distance = sequenceDistance(sequence, cumulativeAck);
  return distance === 0 || distance < 0x80000000;
}

/**
 * A transport-independent buffer for server-authoritative client prediction.
 *
 * Inputs are retained in uint32 sequence order. A cumulative server ACK drops
 * confirmed inputs, then reconcile replays the remaining inputs over the
 * supplied authoritative state. The core knows nothing about movement or
 * physics: callers own state/input cloning and the deterministic replay
 * callback. If inputs are mutable, clone them before push.
 */
export class PredictionBuffer<TInput> {
  readonly maxPending: number;
  private entries: PendingInput<TInput>[] = [];
  private newestSequence: number | null = null;
  private lastAckSequence: number | null = null;
  private overflowCountValue = 0;

  constructor(config: PredictionBufferConfig = {}) {
    const configured = config.maxPending ?? DEFAULT_MAX_PENDING_INPUTS;
    if (!Number.isSafeInteger(configured) || configured < 1) {
      throw new RangeError("maxPending must be a positive safe integer");
    }
    this.maxPending = configured;
  }

  get pendingCount(): number {
    return this.entries.length;
  }

  get overflowCount(): number {
    return this.overflowCountValue;
  }

  get lastAcknowledgedSequence(): number | null {
    return this.lastAckSequence;
  }

  /** A shallow snapshot; input values remain caller-owned. */
  get pendingInputs(): readonly PendingInput<TInput>[] {
    return this.entries.slice();
  }

  push(seq: number, input: TInput): PushInputResult<TInput> {
    const sequence = seq >>> 0;
    if (
      (this.lastAckSequence !== null &&
        !isForwardSequence(this.lastAckSequence, sequence)) ||
      (this.newestSequence !== null &&
        !isForwardSequence(this.newestSequence, sequence))
    ) {
      return { accepted: false, reason: "duplicate-or-stale" };
    }

    this.entries.push({ seq: sequence, input });
    this.newestSequence = sequence;

    if (this.entries.length <= this.maxPending) {
      return { accepted: true };
    }

    const dropped = this.entries.shift()!;
    this.overflowCountValue++;
    return { accepted: true, dropped };
  }

  /** Apply a wrap-safe cumulative ACK. Duplicate/stale ACKs are no-ops. */
  acknowledge(ackSeq: number): AcknowledgeResult {
    const ack = ackSeq >>> 0;
    if (
      this.lastAckSequence !== null &&
      ack !== this.lastAckSequence &&
      !isForwardSequence(this.lastAckSequence, ack)
    ) {
      return {
        advanced: false,
        acknowledgedCount: 0,
        pendingCount: this.entries.length,
      };
    }
    if (ack === this.lastAckSequence) {
      return {
        advanced: false,
        acknowledgedCount: 0,
        pendingCount: this.entries.length,
      };
    }

    this.lastAckSequence = ack;
    let acknowledgedCount = 0;
    while (
      this.entries.length > 0 &&
      isAcknowledgedBy(this.entries[0].seq, ack)
    ) {
      this.entries.shift();
      acknowledgedCount++;
    }

    return {
      advanced: true,
      acknowledgedCount,
      pendingCount: this.entries.length,
    };
  }

  /**
   * Replay every pending input over an authoritative state WITHOUT
   * acknowledging. Use when a correction arrives out of band from the ACK
   * frontier — reconcile is the usual entry point.
   */
  replay<TState>(
    authoritativeState: TState,
    replay: (state: TState, input: TInput, seq: number) => TState,
  ): TState {
    let state = authoritativeState;
    for (const pending of this.entries) {
      state = replay(state, pending.input, pending.seq);
    }
    return state;
  }

  reconcile<TState>(
    authoritativeState: TState,
    ackSeq: number,
    replay: (state: TState, input: TInput, seq: number) => TState,
  ): ReconciliationResult<TState> {
    const acknowledgement = this.acknowledge(ackSeq);
    let state = authoritativeState;
    for (const pending of this.entries) {
      state = replay(state, pending.input, pending.seq);
    }
    return {
      state,
      pendingCount: this.entries.length,
      acknowledgedCount: acknowledgement.acknowledgedCount,
    };
  }

  /**
   * Clear pending prediction state at a connection/session boundary. Pass the
   * authoritative cumulative ACK when known to seed subsequent sequence order.
   */
  reset(ackSeq?: number): void {
    this.entries.length = 0;
    this.newestSequence = null;
    this.lastAckSequence = ackSeq === undefined ? null : ackSeq >>> 0;
    this.overflowCountValue = 0;
  }
}

/**
 * Callback-based helpers for consumers that want to measure or visually blend
 * a correction without coupling this core to a vector or math package.
 *
 * Mirrors PredictionCorrection in csharp/Mmokit.Sdk.Core/PredictionBuffer.cs so
 * the two SDK cores present one contract.
 */
export const predictionCorrection = {
  measure<TState>(
    predictedState: TState,
    reconciledState: TState,
    measureError: (predicted: TState, reconciled: TState) => number,
  ): number {
    return measureError(predictedState, reconciledState);
  },

  /** factor is clamped to 0..1 before it reaches the blend callback. */
  blend<TState>(
    displayedState: TState,
    reconciledState: TState,
    factor: number,
    blend: (displayed: TState, reconciled: TState, factor: number) => TState,
  ): TState {
    const clamped = Math.max(0, Math.min(1, factor));
    return blend(displayedState, reconciledState, clamped);
  },
};

File: csharp/Mmokit.Sdk.Core/PredictionBuffer.cs
using System;
using System.Collections.Generic;

namespace Mmokit.Sdk.Core
{
    /// One locally-predicted input that has not yet been acknowledged by the
    /// authoritative server.
    public readonly struct PendingInput<TInput>
    {
        public PendingInput(uint sequence, TInput input)
        {
            Sequence = sequence;
            Input = input;
        }

        public uint Sequence { get; }
        public TInput Input { get; }
    }

    /// <summary>
    /// Bounded, engine-agnostic input history for client prediction.
    ///
    /// The buffer does not simulate game state and never decides authority. A
    /// client records each input after assigning its wire sequence, then feeds
    /// authoritative acknowledgements and snapshots back into this class.
    /// Reconciliation always starts from the authoritative state and replays
    /// only the still-unacknowledged inputs through a caller-owned callback.
    ///
    /// Sequence comparisons use RFC 1982-style uint serial arithmetic, so
    /// acknowledgement and replay remain correct across uint wraparound.
    /// Public operations are thread-safe so network acknowledgements may arrive
    /// on a background callback while the game thread records or replays input.
    /// Caller replay callbacks are always invoked after releasing the buffer
    /// lock.
    /// </summary>
    public sealed class PredictionBuffer<TInput>
    {
        public const int DefaultCapacity = 128;

        readonly object _sync = new();
        readonly List<PendingInput<TInput>> _pending;
        bool _hasLastAcknowledgement;
        uint _lastAcknowledgedSequence;
        long _droppedInputCount;

        public PredictionBuffer(int capacity = DefaultCapacity)
        {
            if (capacity <= 0)
                throw new ArgumentOutOfRangeException(nameof(capacity), "Prediction capacity must be positive.");

            Capacity = capacity;
            _pending = new List<PendingInput<TInput>>(capacity);
        }

        public int Capacity { get; }
        public int Count
        {
            get
            {
                lock (_sync)
                    return _pending.Count;
            }
        }

        /// A stable snapshot of the pending history. Mutating the buffer after
        /// reading this property cannot change the returned collection.
        public IReadOnlyList<PendingInput<TInput>> Pending
        {
            get
            {
                lock (_sync)
                    return _pending.ToArray();
            }
        }

        /// Number of oldest, still-unacknowledged inputs evicted to honor
        /// Capacity. A non-zero value is useful telemetry and normally means a
        /// consumer should prefer a hard correction over a long visual blend.
        public long DroppedInputCount
        {
            get
            {
                lock (_sync)
                    return _droppedInputCount;
            }
        }

        public bool HasLastAcknowledgement
        {
            get
            {
                lock (_sync)
                    return _hasLastAcknowledgement;
            }
        }

        public uint LastAcknowledgedSequence
        {
            get
            {
                lock (_sync)
                {
                    if (!_hasLastAcknowledgement)
                        throw new InvalidOperationException("No input acknowledgement has been observed.");
                    return _lastAcknowledgedSequence;
                }
            }
        }

        /// True when candidate is later than reference in the uint serial
        /// number space. Comparisons exactly half the uint range apart are
        /// intentionally treated as unordered.
        public static bool IsSequenceNewer(uint candidate, uint reference)
        {
            uint distance = unchecked(candidate - reference);
            return distance != 0 && distance < 0x80000000u;
        }

        /// Record a newly-sent local input. Returns false for a duplicate,
        /// out-of-order sequence, or an input already covered by the latest
        /// acknowledgement. When full, the oldest pending input is evicted.
        public bool TryRecord(uint sequence, TInput input)
        {
            lock (_sync)
            {
                if (_hasLastAcknowledgement &&
                    !IsSequenceNewer(sequence, _lastAcknowledgedSequence))
                {
                    return false;
                }

                if (_pending.Count > 0 && !IsSequenceNewer(sequence, _pending[_pending.Count - 1].Sequence))
                    return false;

                if (_pending.Count == Capacity)
                {
                    _pending.RemoveAt(0);
                    _droppedInputCount++;
                }

                _pending.Add(new PendingInput<TInput>(sequence, input));
                return true;
            }
        }

        /// Apply an authoritative cumulative acknowledgement. Returns the
        /// number of pending inputs removed. Duplicate and stale
        /// acknowledgements are ignored and never move the acknowledgement
        /// frontier backward.
        public int Acknowledge(uint acknowledgedSequence)
        {
            lock (_sync)
                return AcknowledgeLocked(acknowledgedSequence);
        }

        int AcknowledgeLocked(uint acknowledgedSequence)
        {
            if (_hasLastAcknowledgement && !IsSequenceNewer(acknowledgedSequence, _lastAcknowledgedSequence))
                return 0;

            _hasLastAcknowledgement = true;
            _lastAcknowledgedSequence = acknowledgedSequence;

            int removeCount = 0;
            while (removeCount < _pending.Count)
            {
                uint sequence = _pending[removeCount].Sequence;
                if (sequence != acknowledgedSequence && !IsSequenceNewer(acknowledgedSequence, sequence))
                    break;
                removeCount++;
            }

            if (removeCount > 0)
                _pending.RemoveRange(0, removeCount);
            return removeCount;
        }

        /// Replay all pending inputs, in sequence order, starting from an
        /// authoritative state. The callback owns all simulation semantics.
        public TState Replay<TState>(
            TState authoritativeState,
            Func<TState, uint, TInput, TState> replayInput)
        {
            if (replayInput == null)
                throw new ArgumentNullException(nameof(replayInput));

            PendingInput<TInput>[] pending;
            lock (_sync)
                pending = _pending.ToArray();

            TState state = authoritativeState;
            for (int i = 0; i < pending.Length; i++)
            {
                PendingInput<TInput> input = pending[i];
                state = replayInput(state, input.Sequence, input.Input);
            }
            return state;
        }

        /// Acknowledge through acknowledgedSequence, then rebuild predicted
        /// state from the supplied authoritative snapshot and remaining input
        /// history.
        public TState Reconcile<TState>(
            uint acknowledgedSequence,
            TState authoritativeState,
            Func<TState, uint, TInput, TState> replayInput)
        {
            if (replayInput == null)
                throw new ArgumentNullException(nameof(replayInput));

            PendingInput<TInput>[] pending;
            lock (_sync)
            {
                AcknowledgeLocked(acknowledgedSequence);
                pending = _pending.ToArray();
            }

            TState state = authoritativeState;
            for (int i = 0; i < pending.Length; i++)
            {
                PendingInput<TInput> input = pending[i];
                state = replayInput(state, input.Sequence, input.Input);
            }
            return state;
        }

        /// Drop all pending history and the acknowledgement frontier. This is
        /// appropriate after reconnect or another explicit session reset.
        public void Reset()
        {
            lock (_sync)
            {
                _pending.Clear();
                _hasLastAcknowledgement = false;
                _lastAcknowledgedSequence = 0;
                _droppedInputCount = 0;
            }
        }
    }

    /// Callback-based helpers for consumers that want to measure or visually
    /// blend a correction without coupling the SDK to a vector/math package.
    public static class PredictionCorrection
    {
        public static double Measure<TState>(
            TState predictedState,
            TState reconciledState,
            Func<TState, TState, double> measureError)
        {
            if (measureError == null)
                throw new ArgumentNullException(nameof(measureError));
            return measureError(predictedState, reconciledState);
        }

        public static TState Blend<TState>(
            TState displayedState,
            TState reconciledState,
            double factor,
            Func<TState, TState, double, TState> blend)
        {
            if (blend == null)
                throw new ArgumentNullException(nameof(blend));

            double clamped = Math.Max(0.0, Math.Min(1.0, factor));
            return blend(displayedState, reconciledState, clamped);
        }
    }
}
